test/fixtures/s3-fixture/src/main/java/fixture/s3/S3HttpHandler.java

```diff
@@ -543,18 +472,162 @@ MultipartUpload getUpload(String uploadId) {
+    }
+
+    public class S3Request {
+        private final String method;
+        private final String path;
+        private final Map<String, List<String>> queryParameters;
 
-        public String request() {
-            return method + " " + uri();
+        public S3Request(String method, String path, Map<String, List<String>> queryParameters) {
+            this.method = method;
+            this.path = path;
+            this.queryParameters = queryParameters;
         }
 
-        public String uri() {
-            if (query.isEmpty()) {
-                return path;
-            } else {
-                return path + "?" + query;
+        public String method() {
+            return method;
+        }
+
+        public String path() {
+            return path;
+        }
+
+        public Map<String, List<String>> queryParameters() {
+            return queryParameters;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (obj == this) return true;
+            if (obj == null || obj.getClass() != this.getClass()) return false;
+            var that = (S3Request) obj;
+            return Objects.equals(this.method, that.method)
+                && Objects.equals(this.path, that.path)
+                && Objects.equals(this.queryParameters, that.queryParameters);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(method, path, queryParameters);
+        }
+
+        @Override
+        public String toString() {
+            return Strings.format("RequestComponents[method=%s, path=%s, queryParameters=%s]", method, path, queryParameters);
+        }
+
+        public boolean hasQueryParamOnce(String name) {
+            final var values = queryParameters.get(name);
+            return values != null && values.size() == 1;
+        }
+
+        public String getQueryParamOnce(String name) {
+            final var values = queryParameters.get(name);
+            assertNotNull(name, values);
+            assertEquals(name + "=" + values, 1, values.size());
+            return values.get(0);
+        }
+
+        public Optional<String> getOptionalQueryParam(String name) {
+            final var values = queryParameters.get(name);
+            if (values == null) {
+                return Optional.empty();
             }
+            assertEquals(name + "=" + values, 1, values.size());
+            return Optional.of(values.get(0));
+        }
+
+        private boolean isBucketRootPath() {
+            return path.equals("/" + bucket) || path.equals("/" + bucket + "/");
+        }
+
+        private boolean isUnderBucketRootAndBasePath() {
+            return path.startsWith("/" + bucketAndBasePath + "/");
+        }
+
+        public boolean isHeadObjectRequest() {
+            return "HEAD".equals(method) && path.startsWith("/" + S3HttpHandler.this.bucketAndBasePath + "/");
```

Comment:
Is `path.startsWith("/" + S3HttpHandler.this.bucketAndBasePath + "/")` the same as `isUnderBucketRootAndBasePath()`?

Reply:
Ah yeah missed that one